### ISSUE TITLE
Compiler doesn't detect out-of-bounds error when accessing a struct member through a reference

### ISSUE THREAD

**Issue by Phillip-Bociu** [rust-lang/rust#99296]
I tried this code:

```rust
struct Foo {
    arr: [u8; 5]
}

fn main() {
    let foo = Foo{ arr: [0; 5] };
    let bar = &foo;
    println!("{}", bar.arr[100]);
}
```

I expected to see this happen: 
compiler error because of the out-of-bounds error

Instead, this happened: 
code compiled without a single error

### Meta

`rustc --version --verbose`:
```
rustc 1.62.0 (a8314ef7d 2022-06-27)
binary: rustc
commit-hash: a8314ef7d0ec7b75c336af2c9857bfaf43002bfc
commit-date: 2022-06-27
host: x86_64-pc-windows-msvc
release: 1.62.0
LLVM version: 14.0.5
```

<details><summary>Backtrace</summary>
<p>
</p>
</details>

**Comment by asquared31415** [comment]
```rust
fn main() {
    let arr: &[u8; 5] = &[0, 0, 0, 0, 0];
    let value = arr[100];
    println!("{}", value);
}
```
Even simpler code, it doesn't need a struct at all.

**Comment by asquared31415** [comment]
I believe this happens because there is no `Index` for `&[T; N]` and it instead coerces to a `&[T]` slice reference first, which we cannot do the bounds check on, since it has dynamic length.
